File: sort/c/src/sort/insertionSort.c
// sort/c/src/sort/insertionSort.c

#include <stdio.h>
#include <stdbool.h>
#include "sort.h"

/*
------------------------------------------------------------------------
    insertionSort  -   insertion sort implementation
------------------------------------------------------------------------
*/

// Insertion sort function.
void insertionSort(array *arptr, bool rev) {
    if (arptr->len > 1){
        for (size_t i = 0; i < arptr->len; i++) {
            size_t j = i;
            while (j > 0 && ((!rev && arptr->ar[j - 1] > arptr->ar[j]) || (rev && arptr->ar[j - 1] < arptr->ar[j]))){
                swap(arptr, j, j - 1);
                j -= 1;
            }
        }
    }
}

File: sort/go/src/sort/insertionSort.go
// sort/go/src/sort/insertionSort.go

package sort

// ------------------------------------------------------------------------
//     insertionSort  -   insertion sort implementation
// ------------------------------------------------------------------------

func InsertionSort(ar *[]float32, rev bool) {
	var j int
	for i := range *ar {
		j = i
		for j > 0 && ((!rev && (*ar)[j-1] > (*ar)[j]) || (rev && (*ar)[j-1] < (*ar)[j])) {
			swap(ar, j, j-1)
			j -= 1
		}
	}
}
